### ISSUE TITLE
Upcoming Closure Compiler change will break  some *_typecheck targets

### ISSUE THREAD

**Issue by lauraharker** [bazelbuild/rules_closure#426]
I'm working on a change to the Closure Compiler that will make referencing Closure dependency methods like `goog.provide` or `goog.module` an error, unless `goog` is defined somewhere. Right now the compiler (unintentionally) allows this pattern.

So any targets using `goog.{provide,module}` that don't include base.js will no longer compile.

I ran into a breakage while testing this inside Google in one of the rules_closure tests (`closure/compiler/test/linting:lenient_love`):
```
  closure/compiler/test/linting/love.js:15: ERROR - variable goog is undeclared
  goog.provide('Love');
  ^
```

From what I can tell, library-level typechecking does not pull in a definition of base.js, even if `no_closure_library` is false. (Is that right?) So anyone using library-level typechecking and the Closure primitives will start seeing this 'undefined goog' error.

Tiago from the build team has suggested a workaround where all library-level typechecking targets get a dummy base.js.i.js file, which just defines goog. (we had trouble getting the aspect to actually properly run over base.js, hence the dummy base.js.i.js). I have a PR for that workaround that I'll send today.

I could also unblock submitting the Closure Compiler change inside google by adding base.js to the `closure/compiler/test/linting:lenient_love` srcs. Any non-Google users of rules_closure might still be broken once this change is in the Closure Compiler release, but it could be a temporary workaround.

**Comment by gkdn** [comment]
I think by default all libraries get the closure dep. See here:
https://github.com/bazelbuild/rules_closure/blob/74dc51fa78e19f3a9fb7cbefbc7969508e4b8b6a/closure/private/defs.bzl#L108

**Comment by lauraharker** [comment]
Either some or all library-level typechecking builds do not get the closure base dep. Hence the 'undefined goog' error.

I admit I'm not familiar enough with rules_closure to know if using the library-level _typecheck targets is common or not, though.

The result of the `collect_js` method call you linked that adds closure base is stored in a variable `js`: https://github.com/bazelbuild/rules_closure/blob/74dc51fa78e19f3a9fb7cbefbc7969508e4b8b6a/closure/compiler/closure_js_library.bzl#L162

and then the updated sources from that call, which include closure_library_base, are passed in the exported provider: https://github.com/bazelbuild/rules_closure/blob/74dc51fa78e19f3a9fb7cbefbc7969508e4b8b6a/closure/compiler/closure_js_library.bzl#L351

However, the`library_level_checks` call doesn't receive the updated sources, just the original, potentially closure base-less sources: https://github.com/bazelbuild/rules_closure/blob/74dc51fa78e19f3a9fb7cbefbc7969508e4b8b6a/closure/compiler/closure_js_library.bzl#L298

So maybe the solution is to switch https://github.com/bazelbuild/rules_closure/blob/74dc51fa78e19f3a9fb7cbefbc7969508e4b8b6a/closure/compiler/closure_js_library.bzl#L298 to take `js.srcs`, not `srcs`?

**Comment by gkdn** [comment]
Looks at it, yes it seems like it is getting only the original srcs. Passing js.srcs doesn't seem like a good idea because it looks like it has the transitive srcs as well.

The better solution seems like having your fake base.i.js added to js.ijs list when the js is collected. However wouldn't we want a more i.js file there? I'm not sure about the implications about the fake file. 

How do you handle base.js in google3 wrt ijs and type checking? Maybe the rule should be updated to follow the same design?

**Comment by lauraharker** [comment]
Yeah, I agree that actually generating the .i.js is a better solution. @tjgq and I just could not figure out how to get it working after trying for an afternoon, so we tried finding a simpler workaround :| I have a copy of what progress we did make, and I'll see if any of that would be useful as a starting point.

In google3, the non-bazel library-level typechecking rules don't special-case base.js. It has to be explicitly in the deps, and we just produce .ijs the same way that we do for other dependencies
